### PR TITLE
May mon(string comparison)

### PR DESCRIPTION
**Pull request by MayMon-T3807**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by PyaeLinn01 — APPROVED** [review]

File: solutions/string_sort.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+"""
+This module provides a function for sorting the characters in a string.
```
Comment (PyaeLinn01) [inline solutions/string_sort.py:2]: The module docstring mentions the functionality well but could be more concise. Consider removing redundancy, such as repeating "sorting the characters in a string

File: solutions/string_sort.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    Returns:
+        str: A string containing the characters of `input_string`
+             sorted alphabetically.
+
+    Raises:
+        TypeError: If the input is not a string.
+        ValueError: If the string is empty.
+
+    Examples:
```
Comment (PyaeLinn01) [inline solutions/string_sort.py:26]: While the examples are good, adding a mixed-case example (e.g., "aBc") and a string with symbols (e.g., "@hello") would make the examples more robust.

File: solutions/string_sort.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        >>> string_sort("hello")
+        'ehllo'
+        >>> string_sort("Python")
+        'Phnoty'
+        >>> string_sort("12345")
+        '12345'
+    """
+    if not isinstance(input_string, str):
```
Comment (PyaeLinn01) [inline solutions/string_sort.py:34]: The error messages are clear, but consider adding examples of valid and invalid inputs in the docstring Raises section for better user understanding.

File: solutions/string_sort.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        raise TypeError("Input must be a string.")
+    if not input_string:
+        raise ValueError("Input string cannot be empty.")
+
+    return "".join(sorted(input_string))
```
Comment (PyaeLinn01) [inline solutions/string_sort.py:39]: This would sort without distinguishing between uppercase and lowercase letters.

File: solutions/tests/test_string_sort.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+- Normal functionality
+- Defensive assertions for invalid inputs
+- Boundary cases
+
+@author: May Mon Thant
+"""
+
+from solutions.string_sort import string_sort
+import unittest
+import sys
```
Comment (PyaeLinn01) [inline solutions/tests/test_string_sort.py:14]: Using sys.path.append for imports might not be necessary if the project structure is well-defined, and the module is properly installed or included in PYTHONPATH. Consider using relative imports if applicable.

File: solutions/tests/test_string_sort.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+from pathlib import Path
+
+sys.path.append(str(Path(__file__).resolve().parents[2]))
+
+
+# noqa: E402
+class TestStringSort(unittest.TestCase):
```
Comment (PyaeLinn01) [inline solutions/tests/test_string_sort.py:21]: The class name is clear and follows PascalCase convention. The docstring is appropriate. No changes needed here.

File: solutions/tests/test_string_sort.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        """Test sorting an empty string raises ValueError."""
+        with self.assertRaises(ValueError):
+            string_sort("")
+
+    def test_non_string_input(self):
+        """Test passing a non-string input raises TypeError."""
+        with self.assertRaises(TypeError):
+            string_sort(123)
+
+    def test_single_character(self):
```
Comment (PyaeLinn01) [inline solutions/tests/test_string_sort.py:50]: These tests are good examples of boundary cases. Consider adding more such cases, like:
An already sorted string (e.g., "abc").
A string with repeated characters (e.g., "aabb").

File: solutions/tests/test_string_sort.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        """Test sorting a single character string."""
+        self.assertEqual(string_sort("a"), "a")
+
+    def test_boundary_case_two_characters(self):
+        """Test sorting a string with two characters."""
+        self.assertEqual(string_sort("ba"), "ab")
+
```
Comment (PyaeLinn01) [inline solutions/tests/test_string_sort.py:57]: Consider testing strings with mixed content (letters, numbers, special characters) to ensure consistent behavior:

def test_mixed_content_string(self):
    """Test sorting a string with mixed content."""
    self.assertEqual(string_sort("a1b2!c"), "!12abc")
